src/lib/Simulator.ts

```diff
@@ -0,0 +1,221 @@
+    private currentState: WorldState
+    private history: WorldState[]
+
+    constructor(initialState: WorldSetup) {
+        this.initialState = initialState
+        this.daysPerTurn = 10
+        this.scaleFactor = initialState.gdpPerDay * 0.2
+        this.currentState = this.computeInitialWorldState()
+        this.history = []
+    }
+
+    /**
+     * Processes
+     */
+    nextTurn(actionsInTurn: PlayerActions): NexTurnState {
+
+        // TODO: add guards to guarantee that player actions and capability improvements match the expectations.
+        // Assumption: The frontend returns ALL the player actions that are enabled and that should be factored in.
+
+        // Create a new copy of the current state to avoid side effects that can pollute the history
+        const nextStateCandidate = this.clone(this.currentState)
+
+        // Factor in the recurring effects of player actions.
+        for (const containmentPolicy of actionsInTurn.containmentPolicies) {
+            nextStateCandidate.indicators = containmentPolicy.recurringEffect(nextStateCandidate)
+        }
+
+        for (const capabilityImprovement of actionsInTurn.capabilityImprovements) {
+            nextStateCandidate.indicators = capabilityImprovement.recurringEffect(nextStateCandidate)
+        }
+
+        // Factor in any new player actions and random events.
+        const newPlayerActions: ContainmentPolicy[] = this.findNewContainmentPolicies(actionsInTurn.containmentPolicies)
+        for (const containmentPolicy of newPlayerActions) {
+            nextStateCandidate.indicators = containmentPolicy.immediateEffect(nextStateCandidate)
+        }
+
+        const newCapabilities: CapabilityImprovements[] = this.findNewCapabilities(actionsInTurn.capabilityImprovements)
+        for (const capability of newCapabilities) {
+            actionsInTurn
+            nextStateCandidate.indicators = capability.immediateEffect(nextStateCandidate)
+        }
+
+        // Add any new random events that happen
+        const newRandomEvents = this.pickRandomEvents(nextStateCandidate)
+        for (const randomEvent of newRandomEvents) {
+            nextStateCandidate.indicators = randomEvent.immediateEffect(nextStateCandidate)
+        }
+
+        // Save the candidate state as the new current state
+        this.commitState(this.computeNextWorldState(nextStateCandidate), actionsInTurn, newRandomEvents)
+
+        return this.clone({
+            newRandomEvents: newRandomEvents,
+            currentState: this.currentState
+        });
+    }
+
+    /**
+     * Allows the caller to obtain a snapshot of the current simulator state.
+     */
+    state(): SimulatorState {
+        const simulatorStateSnapshot: SimulatorState = {
+            initialState: this.initialState,
+            currentState: this.currentState,
+            history: this.history
+        }
+        // Return an immutable copy of the state.
+        return this.clone(simulatorStateSnapshot)
```

Comment:
I'm using this clone function to provide immutable copies of the internal state of the simulator to the outside world. That way if the gameplay logic for some reason changes anything in this object, it will not affect the internal state of the simulator. AKA defensive copying